Python/specialize.c

```diff
@@ -1195,3 +1199,33 @@ _Py_Specialize_BinaryAdd(PyObject *left, PyObject *right, _Py_CODEUNIT *instr)
     assert(!PyErr_Occurred());
     return 0;
 }
+
+int
+_Py_Specialize_BinaryMultiply(PyObject *left, PyObject *right, _Py_CODEUNIT *instr)
+{
+    PyTypeObject *left_type = Py_TYPE(left);
+    if (left_type != Py_TYPE(right)) {
+        SPECIALIZATION_FAIL(BINARY_MULTIPLY, SPEC_FAIL_DIFFERENT_TYPES);
```

Comment:
It might be worth making this more specific to find what else we should specialize. That can come in another PR.

Reply:
This almost certainly isn't worth refining further.

For `BINARY_ADD` differing types caused very few specialization failures.
https://github.com/faster-cpython/ideas/issues/74#issuecomment-898408499

Reply:
Oops, I placed this comment at the wrong location. (I'd meant it for where your new comment is at https://github.com/python/cpython/pull/28727#discussion_r722260167)